lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+                     function_ref<Value(ArrayRef<Value>)> createOp) {
+  // If the result is never used, no point optimizing this. It will
+  // also complicated error handling in getLowestBitAndHigestBitRequired.
+  assert(!narrowingCandidate->getUsers().empty() &&
+         "narrowingCandidate must have at least one use.");
+  assert(narrowingCandidate->getNumResults() == 1 &&
+         "narrowingCandidate must have exactly one result");
+  assert(outerExtractOp &&
+         "narrowOperationWidth must be called from canonicalizing an "
+         "outerExtractOp, which serves as the root op.");
+  IntegerType narrowingCandidateType =
+      narrowingCandidate->getResultTypes().front().cast<IntegerType>();
+
+  size_t highestBitRequired;
+  size_t lowestBitRequired;
+  size_t originalOpWidth = narrowingCandidateType.getIntOrFloatBitWidth();
+
+  std::tie(lowestBitRequired, highestBitRequired) =
+      getLowestBitAndHighestBitRequired(narrowingCandidate, narrowTrailingBits,
+                                        originalOpWidth);
+
+  // Give up, because we can't make it narrower than it already is.
+  if (lowestBitRequired == 0 && highestBitRequired == originalOpWidth - 1)
+    return false;
+
+  auto loc = narrowingCandidate->getLoc();
+  size_t narrowedWidth = highestBitRequired - lowestBitRequired + 1;
+  auto narrowedType = rewriter.getIntegerType(narrowedWidth);
+  Value narrowedOperation =
+      createOp(SmallVector<Value>(llvm::map_range(inputs, [&](auto input) {
+        return rewriter.create<ExtractOp>(loc, narrowedType, input,
+                                          lowestBitRequired);
+      })));
+
+  auto getNarrowedExtractReplacement = [&](ExtractOp extractUse) -> Value {
+    auto oldLowBit = extractUse.lowBit();
+
+    assert(oldLowBit >= lowestBitRequired &&
+           "incorrectly deduced the lowest bit required in usage arguments.");
+
+    auto loc = extractUse.getLoc();
+
+    if (narrowedWidth == extractUse.getType().getWidth()) {
+      return narrowedOperation;
+    }
+
+    uint32_t newLowBit = oldLowBit - lowestBitRequired;
+    Value narrowedExtract = rewriter.create<ExtractOp>(
+        loc, extractUse.getType(), narrowedOperation, newLowBit);
+    return narrowedExtract;
+  };
+
+  // Pattern rewriter requires that that the root operation (ie: outerExtractOp)
+  // in question to be updated in place, replaced or erased, so this is
+  // necessary.
+  rewriter.replaceOp(outerExtractOp,
+                     getNarrowedExtractReplacement(outerExtractOp));
+
+  // However, for other use sites, calling rewriter.replaceOp or
+  // replaceOpWithNewOp is potentially memory unsafe, as it requires the
+  // rewriting driver to ensure that dangling references in any internal data
+  // structures are properly cleared out.
+  //
+  // Calling narrowingCandidate->getUsers will not include outerExtractOp, since
+  // rewriter.replaceOp will have removed it from the iterator.
+  for (Operation *user :
```

Comment:
Are there situations where narrowing candidate is used more than once by the same operation? In those cases, the user would appear more than once in the `getUsers` list.

Reply:
Yes, they could appear multiple times. I think the current code will do the correct thing, but will do some unnecessary duplicated work.

FYI, addressing this issue using the following patch doesn't get rid of the seg faults:

```diff
-  for (Operation *user :
-       llvm::make_early_inc_range(narrowingCandidate->getUsers())) {
+  while (!narrowingCandidate->use_empty()) {
+    Operation *user = *narrowingCandidate->user_begin();
     auto extractUser = cast<ExtractOp>(user);
     rewriter.updateRootInPlace(extractUser, [&]() {
-      extractUser.replaceAllUsesWith(
-          getNarrowedExtractReplacement(extractUser));
+        rewriter.replaceOp(extractUser, getNarrowedExtractReplacement(extractUser));
     });
   }
```

Reply:
actually -- i'm mistaken, in this specific instance, it will never appear multiple times. All uses of narrowingCandidate will be an `comb::ExtractOp`, which has only 1 operand.